core/src/pow/types.rs

```diff
@@ -478,3 +514,40 @@ impl BitVec {
 		self.bits[pos / 8] |= 1 << (pos % 8) as u8;
 	}
 }
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use crate::ser::{BinReader, BinWriter, ProtocolVersion};
+	use rand::Rng;
+	use std::io::Cursor;
+
+	#[test]
+	fn test_proof_rw() {
+		for edge_bits in 10..64 {
+			let mut proof = Proof::new(gen_proof(edge_bits as u32));
+			proof.edge_bits = edge_bits;
+			let mut buf = Cursor::new(Vec::new());
+			let mut w = BinWriter::new(&mut buf, ProtocolVersion::local());
+			if let Err(e) = proof.write(&mut w) {
+				panic!("failed to write proof {:?}", e);
+			}
+			buf.set_position(0);
+			let mut r = BinReader::new(&mut buf, ProtocolVersion::local());
+			match Proof::read(&mut r) {
+				Err(e) => panic!("failed to read proof: {:?}", e),
+				Ok(p) => assert_eq!(p, proof),
+			}
+		}
+	}
+
+	fn gen_proof(bits: u32) -> Vec<u64> {
+		let mut rng = rand::thread_rng();
+		let mut v = Vec::with_capacity(42);
+		for _ in 0..42 {
+			v.push(rng.gen_range(u64::pow(2, bits - 1), u64::pow(2, bits)));
```

Comment:
u64::pow(2, 64) should be 0.
does rng.gen_range(1<<63, 0) really work to generate random numbers between 2^63 and 2^64 ?

Reply:
I believe it would cause a panic in Rust, but somehow works here, perhaps because he second point is exclusive. Anyway i added a special handling for 64 bits